### DISCUSSION TITLE
lsp: support for path aliases & auto-import?

### DISCUSSION THREAD

**Discussion by MagicDuck** [denoland/deno#10947] [Q&A] [answered]
hi, I am just in the process of trying out the snazzy deno LSP in an existing javascript project that builds with webpack and was wondering if deno lsp has support for path aliases in `jsconfig.json` (I tried with a `tsconfig.json` as well just in case). Here's my `jsconfig.json`:

```
{
  "compilerOptions": {
    "checkJs": true,
    "strict": false,
    "allowSyntheticDefaultImports": true,
    "module": "es6",
    "esModuleInterop": true,
    "moduleResolution": "node",
    "jsx": "preserve",
    "baseUrl": "./",
    "paths": {
      "actions/*": ["./src/actions/*"],
      "Application/*": ["./src/Application/*"],
      "animations/*": ["./src/animations/*"],
      "components/*": ["./src/components/*"],
      "history/*": ["./node_modules/history/*"],
      "hoc/*": ["./src/hoc/*"],
      "hooks/*": ["./src/hooks/*"],
      "IntlWrapper/*": ["./src/IntlWrapper/*"],
      "pages/*": ["./src/pages/*"],
      "reducers/*": ["./src/reducers/*"],
      "resources/*": ["./src/resources/*"],
      "selectors/*": ["./src/selectors/*"],
      "services/*": ["./src/services/*"],
      "store/*": ["./src/store/*"],
      "utils/*": ["./src/utils/*"],
      "xm-constants/*": ["./src/xm-constants/*"],
      "test/*": ["./test/*"],
      "xm-spark": ["./node_modules/xm-spark/dist/spark.bundle.js"],
      "xm-spark-table": ["./node_modules/xm-spark/dist/sparkTable.bundle.js"],
      "xm-spark-time": ["./node_modules/xm-spark/dist/sparkTime.bundle.js"]
    }
  },
  "include": ["src/**/*.js", "src/**/*.jsx"],
  "exclude": ["node_modules"]
}
```

When I tried it, I got the following (in neovim):
![image](https://user-images.githubusercontent.com/95201/121823044-8e816d80-cc57-11eb-81ee-42631d9c5f8f.png)

Not sure if there is some flag that needs to be set or if this functionality just doesn't exist yet...
Similarly I was wondering about auto-import code action and the like. I am trying to replace tsserver with deno, but I am not sure if I am misunderstanding deno at this point, maybe it doesn't even deal with `node_modules`...

**Comment by kitsonk** [comment, marked as answer]
The config file option `paths` doesn't make sense in a Deno context, because Deno doesn't just "transpile" code, it runs code, and "paths" is intended to indicate where code will be at runtime.

[Import maps](https://deno.land/manual@v1.11.0/linking_to_external_code/import_maps) which Deno supports solve the problem of relocating code at runtime and would be the equivalent feature for Deno and the Deno language server.  While the language server supports import maps in the configuration, we don't have 100% of the features (like auto-completion) fully working with import maps yet.

**Reply by MagicDuck** [reply]
Ah, I see now, thank you for explaining! 
Couple of follow up questions:
1. It looks like import maps could work for me, but there is one issue, which is that the current code base I am working with uses extension-less imports for the most part. Are there any plans to allow support for that in the future?
2. Is there any existing issue I could follow (to get notified) for implementing `Add 'import { Foo } from 'foobar'` code action when cursor position is on symbol `Foo`? This type of functionality saves a lot of time during development and would definitely like to give deno a try again when that is implemented.

Much appreciate your help!

**Reply by kitsonk** [reply]
> which is that the current code base I am working with uses extension-less imports for the most part. Are there any plans to allow support for that in the future?

Extension-less imports can be modelled in using import maps I believe.

> Is there any existing issue I could follow (to get notified) for implementing Add 'import { Foo } from 'foobar' code action when cursor position is on symbol Foo?

It looks like we didn't have a specific issue, so I opened #10951.

**Reply by MagicDuck** [reply]
hmm, the example I have seen here: https://github.com/WICG/import-maps#extension-less-imports only seems to cover the case where you point to one single js file.
It doesn't deal with the case where:
- you need a wildcard so that the path alias becomes a prefix: for example `"actions/*": ["./src/actions/*"],`. Does import maps have support for wildcards like that and it's just not documented?
- the extension could be either `.js` or `.jsx`

Thank you very much for filing that issue!

**Reply by kitsonk** [reply]
> the extension could be either `.js` or `.jsx`

That is why extensionless imports are bad. 😄 

> you need a wildcard so that the path alias becomes a prefix: for example `"actions/*": ["./src/actions/*"],`.

I know some level of wildcards are supported, but I am not the expert.

**Reply by MagicDuck** [reply]
hehe yeah, gotta work with what's there though...
Thanks, will try to play with the wildcards :)
